src/parser.rs

```diff
@@ -0,0 +1,225 @@
+pub enum ItemTypes {
+    _CREATE_DIRECTORY,
+    _CREATE_SUBVOLUME,
+    _CREATE_SUBVOLUME_INHERIT_QUOTA,
+    _CREATE_SUBVOLUME_NEW_QUOTA,
+    _EMPTY_DIRECTORY,
+    _TRUNCATE_DIRECTORY,
+    _CREATE_FIFO,
+    _IGNORE_PATH,
+    _IGNORE_DIRECTORY_PATH,
+    _REMOVE_PATH,
+    _RECURSIVE_REMOVE_PATH,
+    _ADJUST_MODE,
+    RELABEL_PATH,
+    _RECURSIVE_RELABEL_PATH,
+    CREATE_FILE,
+    _TRUNCATE_FILE,
+    // TODO: adds missing cases
+}
+
+#[derive(Debug, PartialEq)]
+pub struct Mode {
+    masked: bool,
+    mode: Permissions,
+}
+
+#[derive(Debug, PartialEq)]
+pub struct Action<'a> {
+    action_type: ItemTypes,
+    path: &'a OsStr,
+    mode: Mode,
+    user: &'a str,
+    group: &'a str,
+    age: &'a str,
+    argument: &'a str,
+    boot_only: bool,
+    append_or_force: bool,
+    allow_failure: bool,
+}
+
+impl TryFrom<char> for ItemTypes {
+    type Error = &'static str;
+
+    fn try_from(type_char: char) -> Result<Self, Self::Error> {
+        match type_char {
+            'f' => Ok(ItemTypes::CREATE_FILE),
+            'z' => Ok(ItemTypes::RELABEL_PATH),
+            other => todo!("Support for type `{}` is missing", other),
+        }
+    }
+}
+
+fn item_type_char(input: &[u8]) -> IResult<&[u8], ItemTypes> {
+    map_res(one_of(VALID_ITEM_TYPES), ItemTypes::try_from)(input)
+}
+
+fn item_type(input: &[u8]) -> IResult<&[u8], (ItemTypes, bool, bool, bool)> {
+    let (input, (i_type, boot_only, append_or_force, allow_failure)) = tuple((
+        item_type_char,
+        opt(char('!')),
+        opt(char('+')),
+        opt(char('-')),
+    ))(input)?;
+    let (input, _) = space1(input)?;
+
+    Ok((
+        input,
+        (
+            i_type,
+            boot_only.is_some(),
+            append_or_force.is_some(),
+            allow_failure.is_some(),
+        ),
+    ))
+}
+
+fn path(input: &[u8]) -> IResult<&[u8], &OsStr> {
+    // NOTE as per this comment in "opentmpfiles" we keep the parsing simple
+    // > Upstream says whitespace is NOT permitted in the Path argument.
+    //https://github.com/OpenRC/opentmpfiles/blob/09a1675f68d8106ba08acfc72a263843dabdb588/tmpfiles.sh#L505
+    let (input, path_bytes) = take_till(|c| (c as char).is_whitespace())(input)?;
+    Ok((input, OsStr::from_bytes(path_bytes)))
+}
+
+fn mode(input: &[u8]) -> IResult<&[u8], Mode> {
+    let (input, masked) = opt(char('~'))(input)?;
+    let (input, mode_digits) = oct_digit1(input)?;
+
+    let octal_permission = btoi_radix(mode_digits, 8).unwrap();
+
+    Ok((input, Mode { masked: masked.is_some(), mode: Permissions::from_mode(octal_permission) }))
+}
+
+fn parse_line(input: &[u8]) -> IResult<&[u8], Action> {
+    let (input, (action_type, boot_only, append_or_force, allow_failure)) = item_type(input)?;
+    let (input, path_os_str) = path(input)?;
+
+    Ok((
+        input,
+        Action {
+            action_type,
+            path: path_os_str,
+            mode: Mode { masked: false, mode: Permissions::from_mode(0o755)},
```

Comment:
Just a possible suggestion: how about `impl Default for Mode`?

Reply:
> If omitted or when set to "-", the default is used: 0755 for directories, 0644 for all other file objects.
Which one should I choose? Maybe I can add 2 constructors to Mode

Reply:
Right! You could also think about a build pattern or something similar.